### PR TITLE
1364 Expose list of access requests in mcbe to mc support

### PR DESCRIPTION
**Pull request by davidgisbey**
### Context

MC support currently hits the db directly for to list the access requests.  This pr creates an endpoint on MCBE for the support app to hit instead

### Changes proposed in this pull request

- endpoint created
- list of 'requested' access requests returned when endpoint is hit

### Guidance to review
The support app still needs to be refactored to hit the mcbe endpoint.

### CODE REVIEW

File: app/controllers/api/v2/access_requests_controller.rb [hunk 1]
```diff
@@ -1,19 +1,28 @@
...
+      before_action :build_access_request, except: :list
+      before_action :build_list_of_access_requests, except: :approve
 
       def approve
         result = AccessRequestApprovalService.call(@access_request)
 
         render status: 200, json: { result: result }
       end
 
+      def list
```
Comment (dankmitchell) [inline app/controllers/api/v2/access_requests_controller.rb]: As this is the collection the rails convention dictates it should be `index`

File: app/controllers/api/v2/access_requests_controller.rb [hunk 2]
```diff
@@ -1,19 +1,28 @@
...
+      end
+
     private
 
       def build_access_request
         @access_request = authorize AccessRequest.requested.find(params[:id])
       end
+
+      def build_list_of_access_requests
+        @list_of_access_requests = authorize AccessRequest.requested.all
```
Comment (dankmitchell) [inline app/controllers/api/v2/access_requests_controller.rb]: This instance variable should be `@access_requests` and you shouldn’t need the `all`

File: config/routes.rb [hunk 3]
```diff
@@ -75,8 +75,9 @@
       resource :sessions
       resources :site_statuses, only: :update
 
-      resources :access_requests, only: :update do
+      resources :access_requests, only: %i[update show] do
```
Comment (dankmitchell) [inline config/routes.rb]: `show is not part of this PR so no need for that here but I would add `index` then remove the line you added below

File: spec/requests/api/v2/access_request_spec.rb [hunk 4]
```diff
@@ -24,6 +24,115 @@
 
   subject { response }
 
+  describe 'GET #list' do
```
Comment (dankmitchell) [inline spec/requests/api/v2/access_request_spec.rb]: list to index

File: spec/requests/api/v2/access_request_spec.rb [hunk 5]
```diff
@@ -24,6 +24,115 @@
...
+          [
+            {
+              "id" => first_access_request.id,
+              "email_address" => "first_user@test.com",
+              "first_name" => "test",
+              "last_name" => "user",
+              "requester_email" => "second_user@test.com",
+              "requester_id" => second_user.id,
+              "status" => "requested",
+              "organisation" => "First Organisation",
```
Comment (dankmitchell) [inline spec/requests/api/v2/access_request_spec.rb]: For most of these values you can pull them from the parent variable I.e. `first_organisation.name`

File: spec/requests/api/v2/access_request_spec.rb [hunk 6]
```diff
@@ -24,6 +24,115 @@
...
+               organisation: second_organisation.name,
+               reason: 'Leaving current role',
+               request_date_utc: '2019-05-05 00:10:48 UTC',
+               status: 'requested')
+      }
+      before do
+        list_access_requests_route
+      end
+
+      it 'JSON has a data section with the correct attributes' do
```
Comment (benilovj) [inline spec/requests/api/v2/access_request_spec.rb]: there should be a spec (this one or a separate one) that tests the filtering down to `requested` in the controller i.e. you'd seed the DB with an actioned access request and ensure that it's not returned in the response.

File: spec/requests/api/v2/access_request_spec.rb [hunk 7]
```diff
@@ -24,6 +24,115 @@
 
   subject { response }
 
+  describe 'GET #list' do
+    let(:list_access_requests_route) do
+      get "/api/v2/access_requests/#{admin_user.id}/list",
```
Comment (benilovj) [inline spec/requests/api/v2/access_request_spec.rb]: Once you make the change as Dan's suggested, this will change to `get "/api/v2/access_requests"`. You don't want the admin user ID in there as they're unnecessary, the admin user details are already in the `Authorisation` header.

File: app/controllers/api/v2/access_requests_controller.rb [hunk 8]
```diff
@@ -1,19 +1,28 @@
...
+        render status: 200, json: @list_of_access_requests
+      end
+
     private
 
       def build_access_request
         @access_request = authorize AccessRequest.requested.find(params[:id])
       end
+
+      def build_list_of_access_requests
```
Comment (benilovj) [inline app/controllers/api/v2/access_requests_controller.rb]: I wouldn't bother extracting this method just yet, as it's only used in one place.

File: app/controllers/api/v2/access_requests_controller.rb (hunk 2 above)
Comment (benilovj) [reply app/controllers/api/v2/access_requests_controller.rb]: I think it'd be better to split this into two lines:

```ruby
authorize AccessRequest
@access_requests = AccessRequest.requested
```

The first line is a slightly different way to do the authorisation because lists of things are authorised slightly differently to individual things. See `API::V2::ProvidersController#index` and `API::V2::CoursesController#index` that follow the pattern of authorising the class, rather than the collection that you get back from ActiveRecord.

**Comment by benilovj** [comment]
A comment regarding the structure of the API:

```ruby
            {
              "id" => first_access_request.id,
              "email_address" => "first_user@test.com",
              "first_name" => "test",
              "last_name" => "user",
              "requester_email" => "second_user@test.com",
              "requester_id" => second_user.id,
              "status" => "requested",
              "organisation" => "First Organisation",
              "reason" => "Need additional support",
              "request_date_utc" => "2019-05-05T00:10:47.000Z"
            },
```
I think it could be made more expressive, and harder to misuse, if `requester` and `recipient` were explicit associations of the access request of type `users`. Requester is already a proper foreign key. Recipient could be fed into the serialiser from the model in this way:

```ruby
class AccessRequest
  ...
  def recipient
    User.new(first_name: first_name, last_name: last_name, email: email_address)
  end
end
```

**Comment by davidgisbey** [comment]
Thanks for the feedback guys.  Will action it tonight/tomorrow.

File: app/controllers/api/v2/access_requests_controller.rb (hunk 1 above)
Comment (davidgisbey) [reply app/controllers/api/v2/access_requests_controller.rb]: Have actioned this. Thanks Dan.

File: app/controllers/api/v2/access_requests_controller.rb (hunk 2 above)
Comment (davidgisbey) [reply app/controllers/api/v2/access_requests_controller.rb]: What's funny about this is i initially had it as @access_requests, but changed it due to Heidar's voice popping into my head saying it wasn't very clear. Have actioned this. Thanks for the feedback.

File: config/routes.rb (hunk 3 above)
Comment (davidgisbey) [reply config/routes.rb]: Done. Thanks.

File: spec/requests/api/v2/access_request_spec.rb (hunk 4 above)
Comment (davidgisbey) [reply spec/requests/api/v2/access_request_spec.rb]: Done. Thanks

File: spec/requests/api/v2/access_request_spec.rb (hunk 7 above)
Comment (davidgisbey) [reply spec/requests/api/v2/access_request_spec.rb]: Done. Thanks for the feedback.

File: spec/requests/api/v2/access_request_spec.rb (hunk 5 above)
Comment (davidgisbey) [reply spec/requests/api/v2/access_request_spec.rb]: Done. Thanks

Comment (davidgisbey) [reply spec/requests/api/v2/access_request_spec.rb]: There is one exception. When JSON parses a datetime into a string it seems to output it in a slightly different format. However, it is the correct time so i've manually made that attribute by inputting the output value in json. 

before 

`'2019-05-05 00:10:48 UTC'`
after
`'2019-05-05T00:10:47.000Z'`
🤷‍♂

File: spec/requests/api/v2/access_request_spec.rb (hunk 6 above)
Comment (davidgisbey) [reply spec/requests/api/v2/access_request_spec.rb]: Added this into the (now) #index req spec test by let!'ing an approved access request prior to hitting the endpoint. It's not being parsed into JSON 👍

**Comment by davidgisbey** [comment]
Regarding your comments on the structure of the API. I'll pick this up tomorrow (today i guess as it's past midnight) as it will involve refactoring the previous access request work. As it looks like a fairly substantial piece of work, is it worth opening a separate PR or are you happy for me to change  approve endpoint work in this PR?

File: app/controllers/api/v2/access_requests_controller.rb (hunk 8 above)
Comment (davidgisbey) [reply app/controllers/api/v2/access_requests_controller.rb]: Cool 👍

File: app/controllers/api/v2/access_requests_controller.rb [hunk 9]
```diff
@@ -1,19 +1,29 @@
...
+      before_action :build_access_request, except: :index
+      before_action :build_access_requests, except: :approve
 
       def approve
         result = AccessRequestApprovalService.call(@access_request)
 
         render status: 200, json: { result: result }
       end
 
+      def index
```
Comment (dankmitchell) [inline app/controllers/api/v2/access_requests_controller.rb]: Jake has mentioned this but I think it was missed.

This method should be consistent with our other controllers and read like the below so you can remove the build method. We only use a `before_action` to DRY the code up when we're repeating something on multiple actions or to make it read easier.

```
    def index
        authorize AccessRequest
        @access_requests = AccessRequest.requested

        render jsonapi: @access_requests
    end
```

File: app/controllers/api/v2/access_requests_controller.rb [hunk 10]
```diff
@@ -1,19 +1,29 @@
 module API
   module V2
     class AccessRequestsController < API::V2::ApplicationController
-      before_action :build_access_request
+      before_action :build_access_request, except: :index
```
Comment (dankmitchell) [inline app/controllers/api/v2/access_requests_controller.rb]: Might be more explicit to say use `only` rather than `except` i.e. `before_action :build_access_request, only: :show`

File: app/controllers/api/v2/access_requests_controller.rb (hunk 9 above)
Comment (davidgisbey) [reply app/controllers/api/v2/access_requests_controller.rb]: Yea sorry i missed that one. Have now actioned it. Thanks Dan.

File: app/controllers/api/v2/access_requests_controller.rb (hunk 10 above)
Comment (davidgisbey) [reply app/controllers/api/v2/access_requests_controller.rb]: Great. Just changed it to 
` before_action :build_access_request, only: :approve`
Thanks Dan.

File: spec/requests/api/v2/access_request_spec.rb [hunk 11]
```diff
@@ -24,6 +24,117 @@
...
+      end
+
+
+      it "should raise an error" do
+        expect { unauthorised_user_route }.to raise_error Pundit::NotAuthorizedError
+      end
+    end
+
+    context 'when authorised' do
+      let(:first_organisation) { create(:organisation, name: 'First Organisation') }
```
Comment (heidar) [inline spec/requests/api/v2/access_request_spec.rb]: Do we need all this setup? Feel like a lot of this can come from the factories instead of being explicitly set.

File: spec/policies/access_request_policy_spec.rb [hunk 12]
```diff
@@ -18,4 +18,20 @@
...
     end
   end
+
+  permissions :index? do
+    let(:access_request) { build(:access_request) }
+
+    context 'non-admin user' do
+      let(:user) { build(:user) }
+
+      it { should_not permit(user, access_request) }
```
Comment (heidar) [inline spec/policies/access_request_policy_spec.rb]: I don't think you need the `access_request` here. You should be able to just do 

`it { should_not permit(user) }`

That's because the `index?` method doesn't need an access request - we only need the user to see if it can get the list of access requests.

Comment (heidar) [reply spec/policies/access_request_policy_spec.rb]: Separately, if the `index?` and `approve?` would be exactly the same, you can do `permissions :approve?, :index? do ...`

See https://github.com/varvet/pundit#policy-specs

Just figured I'd share, I had forgotten myself :D

File: app/models/access_request.rb [hunk 13]
```diff
@@ -24,6 +24,10 @@ class AccessRequest < ApplicationRecord
     declined
   ].freeze
 
+  def recipient
```
Comment (heidar) [inline app/models/access_request.rb]: What is this used for?

File: spec/requests/api/v2/access_request_spec.rb [hunk 14]
```diff
@@ -24,6 +24,79 @@
...
+              "last_name" => second_access_request.recipient.last_name,
+              "requester_email" => second_access_request.requester.email,
+              "requester_id" => second_access_request.requester.id,
+              "organisation" => second_access_request.organisation,
+              "reason" => second_access_request.reason,
+              "request_date_utc" => '2019-05-05T00:10:48.000Z',
+              "status" => second_access_request.status
+            }
+          ]
+       )
```
Comment (misaka) [inline spec/requests/api/v2/access_request_spec.rb]: Is there a reason this is not a JSON API response? I would expect it to be, offhand ...

File: spec/policies/access_request_policy_spec.rb (hunk 12 above)
Comment (misaka) [reply spec/policies/access_request_policy_spec.rb]: > That's because the index? method doesn't need an access request

I think it needs `AccessRequest` since technically we're checking the user against the model class, and not (as you say) an individual access request.

Comment (heidar) [reply spec/policies/access_request_policy_spec.rb]: hmm I thought that was just for the magic in `authorize` to work

File: app/controllers/api/v2/access_requests_controller.rb [hunk 15]
```diff
@@ -1,14 +1,21 @@
...
         result = AccessRequestApprovalService.call(@access_request)
 
         render status: 200, json: { result: result }
       end
 
+      def index
+        authorize AccessRequest
+        @access_requests = AccessRequest.requested
+
+        render status: 200, json: @access_requests
```
Comment (heidar) [inline app/controllers/api/v2/access_requests_controller.rb]: oh yeah, good catch @misaka - this should be `jsonapi: @access_requests`

File: spec/factories/access_requests.rb [hunk 16]
```diff
@@ -22,8 +22,24 @@
     email_address { Faker::Internet.email }
     first_name { Faker::Name.first_name }
     last_name { Faker::Name.last_name }
-    organisation { create(:organisation) }
+    organisation { requester.organisations.first.name }
```
Comment (defong) [inline spec/factories/access_requests.rb]: is meant to be `name` and not `organisation`?

File: spec/policies/access_request_policy_spec.rb (hunk 12 above)
Comment (heidar) [reply spec/policies/access_request_policy_spec.rb]: Just noticed that we alias `index?` to `approve?` so maybe let's test these the way I suggested before:

`permissions :approve?, :index? do ...`

By modifying the test for `approve?` that is. Then we can get rid of this one.

File: spec/factories/access_requests.rb (hunk 16 above)
Comment (davidgisbey) [reply spec/factories/access_requests.rb]: You are correct. Done. Thanks.

File: spec/policies/access_request_policy_spec.rb (hunk 12 above)
Comment (davidgisbey) [reply spec/policies/access_request_policy_spec.rb]: Cool. Done. Thanks guys

Comment (misaka) [reply spec/policies/access_request_policy_spec.rb]: Spoke in-person, makes sense to me assuming `AccessRequest` really isn't needed.
